### PR TITLE
thrift: keep downstream connection on routing error

### PR DESCRIPTION
**Pull request by tkovacs-2**
Commit Message:
Option added to Thrift router to don't close downstream connection in case of routing or upstream connection problem.

Additional Description:
By default the router closes downstream connection if the current request is not routable or the upstream connection is broken. This causes oneway (or async) requests pending on wire or in kernel buffers will be also dropped even if they are routable and their target cluster is perfectly healthy.

Risk Level:
Low

Testing:
Unittest

Docs Changes:
Comment added to the new API field.

Release Notes:

Platform Specific Features:

Fixes [#12836](https://github.com/envoyproxy/envoy/issues/12836)

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @rgs1

**Comment by rgs1** [comment]
cc: @fishcakez

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc [hunk 1]
```diff
@@ -66,7 +67,9 @@ void UpstreamRequest::onPoolFailure(ConnectionPool::PoolFailureReason reason, ab
 
   // Mimic an upstream reset.
   onUpstreamHostSelected(host);
-  onResetStream(reason);
+  if (!onResetStream(reason)) {
+    parent_.continueDecoding();
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:71]: why is this needed?

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc [hunk 2]
```diff
@@ -200,15 +203,16 @@ bool UpstreamRequest::handleUpstreamData(Buffer::Instance& data, bool end_stream
 
 void UpstreamRequest::onEvent(Network::ConnectionEvent event) {
   ASSERT(!response_complete_);
+  bool end_downstream = true;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:206]: nit: `close_downstream`

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc [hunk 3]
```diff
@@ -262,13 +268,8 @@ void UpstreamRequest::onUpstreamHostSelected(Upstream::HostDescriptionConstShare
...
 
-void UpstreamRequest::onResetStream(ConnectionPool::PoolFailureReason reason) {
-  if (metadata_->messageType() == MessageType::Oneway) {
-    // For oneway requests, we should not attempt a response. Reset the downstream to signal
-    // an error.
-    parent_.resetDownstreamConnection();
-    return;
-  }
+bool UpstreamRequest::onResetStream(ConnectionPool::PoolFailureReason reason) {
+  bool end_downstream = true;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:272]: nit: `close_downstream`

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.h [hunk 4]
```diff
@@ -78,6 +78,8 @@ struct UpstreamRequest : public Tcp::ConnectionPool::Callbacks,
   bool charged_response_timing_{false};
   MonotonicTime downstream_request_complete_time_;
   uint64_t response_size_{};
+
+  bool keep_downstream_;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.h:82]: probably `close_downstream_on_error_` per the previous comments.

**Review by rgs1 — CHANGES_REQUESTED** [review]
Thank you, this a nice option to have. I left some nits and questions.

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc (hunk 1 above)
Comment (tkovacs-2) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:71]: The `UpstreamRequest::start` stops the decoding on downstream if no upstream connection is available immediately.
In the `UpstreamRequest::onPoolFailure` if we don't close the downstream then need to call the `continueDecoding` to be able to get the further requests, otherwise the downstream will just hang.

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.h (hunk 4 above)
Comment (tkovacs-2) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.h]: Done

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc (hunk 3 above)
Comment (tkovacs-2) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc]: Done

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc (hunk 2 above)
Comment (tkovacs-2) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:206]: Done

File: api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto [hunk 5]
```diff
@@ -18,4 +18,7 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message Router {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.thrift.router.v2alpha1.Router";
+
+  // Don't close downstream connection in case of routing or upstream connection problem.
+  bool keep_downstream = 1;
```
[the remark this answers is not in the document]
Comment (tkovacs-2) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto]: Done

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc [hunk 6]
```diff
@@ -107,11 +107,13 @@ void ConnectionManager::sendLocalReply(MessageMetadata& metadata, const DirectRe
   Buffer::OwnedImpl buffer;
   const DirectResponse::ResponseType result = response.encode(metadata, *protocol_, buffer);
```
[the remark this answers is not in the document]
Comment (tkovacs-2) [reply source/extensions/filters/network/thrift_proxy/conn_manager.cc]: Yes, for oneway there is no result, I just wanted to have some trace of the failure at least in statics.
But right, for oneway error or succes reply is not possible anyway, so the `encode()` can be also skipped.

Why `AppException` and `EnvoyException` handled differently in `dispatch()`?
`AppException` calls `sendLocalReply` even with empty metadata, while `EnvoyException` just closes the connection if metadata is empty.

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc [hunk 7]
```diff
@@ -262,13 +268,8 @@ void UpstreamRequest::onUpstreamHostSelected(Upstream::HostDescriptionConstShare
   upstream_host_ = host;
 }
 
-void UpstreamRequest::onResetStream(ConnectionPool::PoolFailureReason reason) {
-  if (metadata_->messageType() == MessageType::Oneway) {
```
Comment (tkovacs-2) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:266]: @rgs1
I have completely removed this special handling for oneway here.
This means a small backward incompatibility since outlier detection and stat counting will be done for oneway as well. Is it ok?

Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:266]: I think that's OK (though that might be a narrow view, since we don't use Oneway). 

@fishcakez thoughts?

**Review by rgs1 — COMMENTED** [review]
This LGTM, thanks. Waiting for @fishcakez's comment on the Oneway semantics change before signing off.

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc (hunk 7 above)
Comment (fishcakez) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:266]: Im not sure this is the right behavior as I dont think we should send an TApplicationException on a oneway, or is this hidden inside ThriftProxy that local reply isnt done then?

File: api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto [hunk 8]
```diff
@@ -18,4 +20,7 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message Router {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.thrift.router.v2alpha1.Router";
+
+  // Close downstream connection in case of routing or upstream connection problem.
+  google.protobuf.BoolValue close_downstream_on_upstream_error = 1;
```
Comment (fishcakez) [inline api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:25]: If we avoid closing a connection safely, then should we always do it?

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc (hunk 7 above)
Comment (fishcakez) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:266]: But to answer the specific question. I think its okay from a thrift perspective to set outlier detection failures on oneways for conenction issues as still failed to proxy the request to the endpoint.

File: api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto (hunk 8 above)
Comment (rgs1) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:25]: I think we can switch to on by default later on, once this has been used for a bit.

Comment (rgs1) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:25]: Oh nvm, we are defaulting to always doing it:

```
  bool close_downstream_on_error =
      PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config, close_downstream_on_upstream_error, true);
      ```

**Review by rgs1 — COMMENTED** [review]
LGTM, one final ask: could you add a line in `docs/root/version_history/current.rst` noting this change? Thanks!

**Comment by rgs1** [comment]
> LGTM, one final ask: could you add a line in `docs/root/version_history/current.rst` noting this change? Thanks!

@tkovacs-2 specifically noting the new behavior.

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc (hunk 7 above)
Comment (tkovacs-2) [reply source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:266]: > Im not sure this is the right behavior as I dont think we should send an TApplicationException on a oneway, or is this hidden inside ThriftProxy that local reply isnt done then?

Yes, in case of oneway there will be nothing sent to wire even if `sendLocalReply` is called, see `ConnectionManager::sendLocalReply`.
Actually the original behavior was not consistent for oneway since in `Router::messageBegin` the failing cases sent out the exception, but here the connection was closed directly.

File: api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto (hunk 8 above)
Comment (tkovacs-2) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:25]: > If we can avoid closing a connection safely, then should we always do it?

I just added this flag to ensure backward compatibility. I don't know, maybe someone has built on the close.
Be careful, the `true` here means the original behavor, so close the connection.

**Comment by tkovacs-2** [comment]
> LGTM, one final ask: could you add a line in `docs/root/version_history/current.rst` noting this change? Thanks!

Ok, sure.

**Review by rgs1 — APPROVED** [review]
Thanks!

File: api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto [hunk 9]
```diff
@@ -18,4 +20,7 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message Router {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.thrift.router.v2alpha1.Router";
+
+  // Close downstream connection in case of routing or upstream connection problem. Default: true
```
Comment (zuercher) [inline api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:24]: Should probably leave a pointer to the runtime flag here.

Comment (tkovacs-2) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:24]: I think there is a misunderstanding here. The default is `true` (close the downstream) which is the original behavior. So if this field is not set then there is no change in the working either.
Maybe it is a little unusual to use true as default for such flag, but I made like this according to comment:
https://github.com/envoyproxy/envoy/pull/20756#discussion_r847439358

Comment (rgs1) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:24]: Yeah this is on by default (current behavior) so I think we should be good. This is less controversial than #20236, so current behavior by default should suffice and save us from needing a runtime flag.

Comment (zuercher) [reply api/envoy/extensions/filters/network/thrift_proxy/router/v3/router.proto:24]: I see. Sounds good.

**Review by zuercher — APPROVED** [review]

**Comment by zuercher** [comment]
@rgs1 ok to merge this?

**Comment by rgs1** [comment]
Yes please, let’s merge - thanks .
